libraries/AP_SmartAudio/AP_SmartAudio.cpp

```diff
@@ -0,0 +1,125 @@
+#include "AP_SmartAudio.h"
+#include "AP_SmartAudioDeviceV2_OneWay.h"
+#include <AP_HAL/AP_HAL.h>
+
+#define SMARTAUDIO_BAUDRATE 4800
+
+const AP_Param::GroupInfo AP_SmartAudio::var_info[] = {
+
+    // @Param: PROTOCOL
+    // @DisplayName: SmartAudio protocol version
+    // @Description: 0: None, 3: Version 2.0 One way
```

Comment:
@OXINARF thanks for the such a complete review! I will go through the questions soon but want to clarify about OneWay and autodetect. The problem is SmartAudio physically is a single-wire half-duplex protocol. That means you have options:
1. To somehow configure your MCU to be a half duplex. I don't know if this possible for a wide range of the supported FC
2. To connect SA line of VTX to TX/RX lines on FC using the resistors scheme. I personally didn't tested this yet. 
3. To connect SA line of VTX to TX line on FC, that's why I calling that OneWay. You're not able to get a response but for 90% of the use cases it is not needed :)

Reply:
> To somehow configure your MCU to be a half duplex. I don't know if this possible for a wide range of the supported FC

I'm not sure if you are aware that we do support half-duplex, you just have to set the appropriate SERIALx_OPTIONS ibt. According to code comments, it's supported in all F4 and F7s, so quite a wide range should support it.

> To connect SA line of VTX to TX line on FC, that's why I calling that OneWay. You're not able to get a response but for 90% of the use cases it is not needed :)

Ok, if we want to support that then we can have 4 options:

- 0: auto-detect
- 1: v1
- 2: v2
- 3: v2.1

Then we can rate limit the sending if needed (I think it's quite low rate already but I'm not against further limiting).

Reply:
@OXINARF No, I wasn't aware F3 supports half-duplex. I found the half-duplex option goes to UART implementation, but is there any example or doc about using that mode? Or there is no difference and I can just use .Read() and .Write() as in normal mode?

Reply:
> No, I wasn't aware F3 supports half-duplex.

Did you meant F4? Because we don't support F3.

> I found the half-duplex option goes to UART implementation, but is there any example or doc about using that mode? Or there is no difference and I can just use .Read() and .Write() as in normal mode?

There is nothing different, usage is the same.

Reply:
Yes, I meant F4. Ok thanks, will try that.